src/HeapJobOrderer.ts

```diff
@@ -0,0 +1,156 @@
+import { Heap } from 'heap-js'
+import { Job } from './Job'
+import { JobOrderer } from './JobOrderer'
+import assert = require('assert')
+
+/**
+ * Manages a set of jobs that have to be run.
+ */
+export class HeapJobOrderer implements JobOrderer {
+	private sourcesHeap: Heap<Job>
+
+	private nonSources: Set<Job> = new Set()
+
+	private inProgress: Set<Job> = new Set()
+
+	private jobToDependents: Map<Job, Set<Job>> = new Map<Job, Set<Job>>()
+
+	private jobToCompletedPrereqs: Map<Job, Set<Job>> = new Map<Job, Set<Job>>()
+
+	/**
+	 * Create the JobOrderer and inform it of the jobs to manage.
+	 *
+	 * @throws When one of the jobs has a prerequisite that is not one of the jobs, or is itself.
+	 * @param jobs - The jobs to manage. They must have correctly populated prerequisites fields.
+	 */
+	constructor(rootJobs: Job[]) {
+		// Initialize the sources heap with its sorting function.
+		this.sourcesHeap = new Heap((job1, job2) => {
+			const job1Dependents = this.jobToDependents.get(job1)
+			const job2Dependents = this.jobToDependents.get(job2)
+
+			assert(
+				job1Dependents && job2Dependents,
+				'Provided jobs are missing from map.',
+			)
+
+			return job2Dependents.size - job1Dependents.size
+		})
+
+		// Add all of the root jobs to dependents map.
+		for (const rootJob of rootJobs) {
+			this.jobToDependents.set(rootJob, new Set())
+		}
+
+		const queue: Job[] = rootJobs
+		const seenSources: Set<Job> = new Set() // We use this because checking whether a heap has an element is O(n).
+
+		// BFS traverse dependency graph, starting with root nodes (the ultimate jobs).
+		while (queue.length) {
+			const job = queue.pop()
+			assert(job, 'Queue with nonzero length should have returned object.')
+
+			// Don't process the same node twice.
+			if (!seenSources.has(job) && !this.nonSources.has(job)) {
```

Comment:
Can you explain why there's two disjoint sets (seenSources, nonSources) and a heap (sourcesHeap) here? I.e. why not just have one set for holding all the nodes, not just sources (nice side effect: it would catch duplicates of any kind, so you get cycle detection for free) and another priority queue to store the sources?

Reply:
Initially we were going to do what you were suggesting, but we left it as two disjoint sets to allow an further optimizations that we may want to do with deciding which job. This is an optional feature that we may or may not implement.

Reply:
`nonSources` is a field that I use at other points in the code. I use `seenSources` because I need to a way to check whether I've already seen a source and I don't wanna check the heap because that's O(n).

Also I don't see how turning `nonSources` to `allJobs` would catch cycles.

Reply:
I see. It would help readability to rename things like so:
- sourcesHeap.
- seenSources -> sourcesSet (to show that it's the same data as sourcesHeap, just as a set; a comment stating this would also be good).
- nonSources -> nonSourcesSet (I don't usually like putting the type in the name, but this way you can tell it's like a parallel to sourcesSet).

Any better suggestions for the names? I'm not totally happy with my suggestions here but I do think they need to be made more clear.

> Also I don't see how turning nonSources to allJobs would catch cycles.

Any time you traverse a tree and add all the elements to a set, if you ever try to add an element that's already there, you know that's a cycle.

Reply:
Hmm I also don't like putting the type in the name, which makes me want to change `sourcesHeap`. How's this:

- `sourcesHeap` -> `sources`
- `seenSources` -> `sourcesHelperSet` (with a comment)
- `nonSources`

This way we still have the parallelism between sources and nonSources. This also indicates that `sourcesHelperSet` is a temporary helper variable.

Reply:
Yeah this is even better.